# Notebook 02 — Preprocessing

Pipeline: lowercase → noise removal → tokenise → remove html-junk tokens → POS tag → lemmatise → stopword removal → flag empties → store

The actual text-transformation functions (contraction expansion, cleaning, tokenizing, lemmatizing, stopword removal) live in [`src/preprocessing.py`](../src/preprocessing.py) — deterministic, corpus-independent logic shared with any other notebook that needs the same pipeline (Phase 6/7). This notebook imports them and still runs/explains each step in its own cell, so nothing about *why* each step exists is hidden behind the import.

In [1]:
import sys
sys.path.insert(0, '..')

import pickle
import pandas as pd
import nltk
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

from src.preprocessing import (
    normalize, clean, tokenize_raw, remove_html_junk,
    get_wordnet_pos, lemmatize, get_stop_words, remove_stopwords,
    NEGATION_WORDS, preprocess,
)

DATASET_PATH = "../data/raw/Emotion_classify_Data.csv"
df = pd.read_csv(DATASET_PATH)

# --- Preprocessing configuration (Phase 3/4) ---
# Each flag lets a step be switched off so its actual effect is testable, not assumed.
# Defaults (all True) reproduce the pipeline's original behavior exactly. Phase 7 will
# run the full variant matrix through src.preprocessing.preprocess() using these same flags.
APPLY_LEMMATIZATION = True  # if False: Lemmatized_Tokens = Cleaned_Tokens (lemmatisation skipped)
APPLY_STOPWORDS = True      # if False: filtered_tokens = Lemmatized_Tokens (stopword removal skipped)
PROTECT_NEGATION = True     # only matters when APPLY_STOPWORDS is True: exclude negation words from the stopword set

## 2.0 Drop contradictory duplicates

EDA (Notebook 01) identified 3 comment pairs that appear under two different labels — always `anger` vs `fear`. All 6 rows are dropped before the pipeline runs to prevent any of them leaking across the train/val/test split.

In [2]:
conflict_comments = (
    df.groupby("Comment")["Emotion"]
    .nunique()
    .loc[lambda s: s > 1]
    .index
)
df = df[~df["Comment"].isin(conflict_comments)].reset_index(drop=True)
print(f"Rows after dropping contradictory duplicates: {len(df)}")

Rows after dropping contradictory duplicates: 5931


**Result:** 5,931 rows (6 dropped — the 3 contradictory pairs).

## 2.1 Lowercase + contraction normalization + noise removal

Lowercase before stripping so all patterns are in a consistent case.

**Correction (was wrong before):** the original version of this step assumed the alpha-only regex below was what destroyed contractions like `don't` → `dont`. Checking the raw CSV directly shows that's not what happens — `data/raw/Emotion_classify_Data.csv` arrives **already lowercased and already stripped of all punctuation** (0 rows contain an apostrophe or an uppercase letter). Negation already arrives pre-merged: `dont` (179 rows), `cant` (110), `didnt` (90), `isnt` (11), `wasnt` (37), `couldnt` (35), `shouldnt` (10), `wont` (24), `doesnt` (22), `arent` (10), `hadnt` (9), `hasnt` (2), `havent` (20), `wouldnt` (8).

None of those fused forms match the `negation_words` protection set used in 2.6 (`"not"`, `"no"`, `"never"`, ...) and none match NLTK's stopword list either (which has `"don't"` with an apostrophe, not `"dont"`). They weren't being deleted — but they also weren't being recognised as negation. Each one just became its own low-frequency, un-lemmatised vocabulary token, fragmenting the negation signal instead of consolidating it.

**Fix:** expand contractions — both the apostrophe'd form (`"don't"`, for any future input that isn't pre-cleaned like this CSV) and the already-merged form actually present in this dataset (`"dont"`) — to the same explicit expansion (`"do not"`), *before* the alpha-only strip runs. This happens in a new `Normalized_Comment` column, so the step is inspectable on its own. The result is that `"not"` always ends up as a literal, consistent token regardless of which surface form the negation arrived in — which is what actually makes the `negation_words` protection in 2.6 effective.

A handful of fused forms are deliberately **not** expanded, because collapsing them would collide with an unrelated real word and corrupt meaning rather than preserve it:

| fused form | real-word meaning | contraction meaning |
|---|---|---|
| `ill` | sick | "i will" |
| `hell` | hell | "he will" |
| `well` | well (very common) | "we will" |
| `its` | possessive | "it is" |
| `were` | past tense of "to be" | "we are" |
| `id` | identity/psych term | "i would" |

These are left as plain tokens — a deliberate precision-over-coverage call, not an oversight. If error analysis later shows these are costing the model something, resolving them would need surrounding context (e.g. a bigram check), not a flat dictionary lookup.

The alpha-only regex (`[^a-zA-Z\s]`) still runs after expansion, for the same reason as before — stripping digits and any remaining punctuation/html special characters. Result is stored in `Cleaned_Comment`; the raw `Comment` column is never reassigned, so it stays exactly as loaded.

In [3]:
df["Normalized_Comment"] = df["Comment"].apply(normalize)
df["Cleaned_Comment"] = df["Normalized_Comment"].apply(clean)
df["Cleaned_Comment"]


0       i seriously hate one subject to death but now ...
1                    i am so full of life i feel appalled
2       i sit here to write i start to dig out my feel...
3       i have been really angry with r and i feel lik...
4       i feel suspicious if there is no one outside l...
                              ...                        
5926                   i begun to feel distressed for you
5927    i left feeling annoyed and angry thinking that...
5928    i were to ever get married i d have everything...
5929    i feel reluctant in applying there because i w...
5930    i just wanted to apologize to you because i fe...
Name: Cleaned_Comment, Length: 5931, dtype: str

**Result:** `Comment` unchanged (truly raw, no longer overwritten). `Normalized_Comment` holds the lowercased, contraction-expanded text (e.g. `"i dont feel happy"` → `"i do not feel happy"`). `Cleaned_Comment` strips that down to `[a-zA-Z\s]` only, same as before. Alphabetic html artifacts (`www`, `href`, `src`, `img`, etc.) still survive this step — removed post-tokenisation in 2.3.

## 2.2 Tokenisation

Split `Cleaned_Comment` into a list of word tokens using `nltk.word_tokenize()`. Output is stored in a new `tokens` column — all downstream steps (POS tagging, lemmatisation, stopword removal) operate on this list, never the raw string.

**Correction:** earlier reasoning here claimed `word_tokenize` was what handled contractions — splitting `"don't"` → `["do", "n't"]`. That was never actually exercised by this pipeline, because there was never an apostrophe left for it to split by the time this step runs (see 2.1: the raw data has no apostrophes to begin with). The real handling of contractions is the explicit `CONTRACTIONS` expansion in 2.1, which turns `"dont"`/`"don't"` into two separate words (`"do"`, `"not"`) *before* tokenisation — so by this step, `word_tokenize` just splits on whitespace like any other two-word phrase. `word_tokenize` is still preferred over `.split()` for everything else it handles correctly (punctuation-adjacent tokens, etc.), just not for contractions in this pipeline.

In [4]:
df['tokens'] = df['Cleaned_Comment'].apply(tokenize_raw)
df['tokens'].head(5)

0    [i, seriously, hate, one, subject, to, death, ...
1       [i, am, so, full, of, life, i, feel, appalled]
2    [i, sit, here, to, write, i, start, to, dig, o...
3    [i, have, been, really, angry, with, r, and, i...
4    [i, feel, suspicious, if, there, is, no, one, ...
Name: tokens, dtype: object

## 2.3 Remove html-junk tokens

After tokenisation, filter out any token that is a known html artifact — fragments like `www`, `href`, `src`, `img`, `rel`, `bookmark`, `class` that survive the regex strip because they are purely alphabetic.

Filtering at the token level (rather than with a broader regex pre-tokenisation) keeps the change surgical: only exact matches on a fixed list are removed, with no risk of clobbering real words with similar substrings.

In [5]:
df["Cleaned_Tokens"] = df["tokens"].apply(remove_html_junk)
df["Cleaned_Tokens"].head(5)

0    [i, seriously, hate, one, subject, to, death, ...
1       [i, am, so, full, of, life, i, feel, appalled]
2    [i, sit, here, to, write, i, start, to, dig, o...
3    [i, have, been, really, angry, with, r, and, i...
4    [i, feel, suspicious, if, there, is, no, one, ...
Name: Cleaned_Tokens, dtype: object

**Result:**

## 2.4 POS tagging

POS tagging happens **before** stopword removal — the tagger uses auxiliary verbs and articles as context clues to correctly label surrounding words.
Removing stopwords first degrades tagging accuracy, which in turn degrades lemmatisation.

> `nltk.pos_tag()` returns Penn Treebank tags (`VBG`, `NN`, `JJ`...). These need to be mapped to WordNet's simpler format before lemmatisation — that mapping is the next step.

In [6]:
df['POS_Tags'] = df['Cleaned_Tokens'].apply(lambda x: nltk.pos_tag(x))
df['POS_Tags'].head(5)

0    [(i, NN), (seriously, RB), (hate, VB), (one, C...
1    [(i, NN), (am, VBP), (so, RB), (full, JJ), (of...
2    [(i, JJ), (sit, NN), (here, RB), (to, TO), (wr...
3    [(i, NNS), (have, VBP), (been, VBN), (really, ...
4    [(i, NN), (feel, VBP), (suspicious, JJ), (if, ...
Name: POS_Tags, dtype: object

In [7]:
print(f'Number of unique tokens: {len(df["POS_Tags"].explode().unique())}')

Number of unique tokens: 11409


**Result:**

## 2.5 Lemmatisation

`WordNetLemmatizer` requires WordNet POS tags (`v`, `n`, `a`, `r`), not Penn Treebank tags.
Write a `get_wordnet_pos(tag)` helper that maps the first character of the Penn tag to the WordNet equivalent — default to `n` (noun) for anything that doesn't match.
Then lemmatise each token using its mapped tag.

> This is tag-aware lemmatisation: `"feeling"` as a verb → `"feel"`, as a noun → `"feeling"`. Blind lemmatisation (no tag) defaults everything to noun and gets verbs wrong.

**Known limitation:** `WordNetLemmatizer` correctly collapses regular verb forms (`"feeling"` → `"feel"`) but fails on irregular forms — `"felt"` is left unchanged even when verb-tagged. This is a documented gap in WordNet's lemmatisation dictionary, verified by isolated testing (`lemmatizer.lemmatize("felt", wordnet.VERB)` returns `"felt"`). Irregular forms like `"felt"`, `"was"`, `"went"` will survive as distinct tokens rather than collapsing to their base form.

This step is controlled by the `APPLY_LEMMATIZATION` flag (set in the imports cell) — defaulting to on, but switchable so lemmatisation's actual effect on downstream results can be measured rather than assumed. Phase 4 will run both settings through the full pipeline to compare.

In [8]:
# get_wordnet_pos is imported from src.preprocessing (see the imports cell) rather than
# redefined here. Quick sanity check that the mapping behaves as described above:
for tag in ['JJ', 'VBG', 'VBD', 'NN', 'RB', 'IN']:
    print(f"{tag} -> {get_wordnet_pos(tag)}")

JJ -> a
VBG -> v
VBD -> v
NN -> n
RB -> r
IN -> n


**Result:** Helper and lemmatizer defined. `.apply()` runs in 2.5b after sample verification.

### 2.5a Before/after verification

Before running `.apply()` on all 5,931 rows — confirm the logic works correctly on a small independent test. If there's a bug in `get_wordnet_pos` (a wrong mapping, a missed case) this is where you catch it cheaply, before it silently corrupts the full column.

The spaCy comparison below also runs here: it confirms the irregular-verb limitation (`felt`, `went`, `had`, `was`) is specific to WordNet's dictionary, not a bug in the pipeline logic.

In [9]:
import spacy
nlp = spacy.load("en_core_web_sm")
doc = nlp("i went to school and had an argument with my friend. I was very angry and frustrated.")
for token in doc:
    print(token.text, token.lemma_)

i I
went go
to to
school school
and and
had have
an an
argument argument
with with
my my
friend friend
. .
I I
was be
very very
angry angry
and and
frustrated frustrate
. .


**Result:** spaCy comparison confirmed — irregular verb forms (`felt`, `went`, `had`, `was`) are a WordNet dictionary gap, not a logic bug. Logic verified on a small sample; scaling to full dataset in 2.5b.

### 2.5b Run on full dataset

Sample verified — apply lemmatisation to all 5,931 rows, then confirm the vocabulary reduction and spot-check the key word variants (`feeling`, `feels`, `felt`) across the full column.

In [10]:
if APPLY_LEMMATIZATION:
    df['Lemmatized_Tokens'] = df['Cleaned_Tokens'].apply(lemmatize)
else:
    df['Lemmatized_Tokens'] = df['Cleaned_Tokens']

before = len(df['Cleaned_Tokens'].explode().unique())
after = len(df['Lemmatized_Tokens'].explode().unique())
print(f"Unique tokens before lemmatization: {before:,}")
print(f"Unique tokens after lemmatization:  {after:,}")
print(f"Vocabulary reduction: {(before - after) / before:.0%}")

Unique tokens before lemmatization: 8,931
Unique tokens after lemmatization:  7,273
Vocabulary reduction: 19%


In [11]:
target_words = ["feeling", "feels", "felt"]

mask = df['Cleaned_Tokens'].apply(lambda tokens: any(target_word in tokens for target_word in target_words))

matches = df[mask]
print(f"Rows containing '{target_words}': {len(matches)}")

for idx in matches.index[:20]:
    original_tokens = df.loc[idx, 'Cleaned_Tokens']       
    lemmatized_tokens = df.loc[idx, 'Lemmatized_Tokens']
    
    if "feeling" in original_tokens :
        position = original_tokens.index("feeling")
    elif "feels" in original_tokens :
        position = original_tokens.index("feels")
    elif "felt" in original_tokens :
        position = original_tokens.index("felt") 
    else :
        None
    
    print(f"Row {idx}:")
    print(f"  original word:   {original_tokens[position]}")
    print(f"  became:          {lemmatized_tokens[position]}")

Rows containing '['feeling', 'feels', 'felt']': 2068
Row 7:
  original word:   feeling
  became:          feel
Row 11:
  original word:   feeling
  became:          feel
Row 13:
  original word:   feeling
  became:          feel
Row 18:
  original word:   feeling
  became:          feel
Row 20:
  original word:   feeling
  became:          feeling
Row 22:
  original word:   feeling
  became:          feel
Row 24:
  original word:   feeling
  became:          feeling
Row 25:
  original word:   feeling
  became:          feeling
Row 27:
  original word:   feeling
  became:          feeling
Row 28:
  original word:   feeling
  became:          feel
Row 34:
  original word:   feeling
  became:          feel
Row 36:
  original word:   feeling
  became:          feel
Row 38:
  original word:   felt
  became:          felt
Row 39:
  original word:   felt
  became:          felt
Row 40:
  original word:   feeling
  became:          feel
Row 41:
  original word:   feeling
  became:          fee

In [12]:
before = len(df['POS_Tags'].explode().unique())
after = len(df['Lemmatized_Tokens'].explode().unique())
print(f"Unique tokens before lemmatization: {before}")
print(f"Unique tokens after lemmatization: {after}")

Unique tokens before lemmatization: 11409
Unique tokens after lemmatization: 7273


**Result:** Lemmatisation applied to all 5,931 rows. Vocabulary reduced by 19% (8,931 → 7,273 unique tokens — slightly lower than before the 2.1 contraction fix, since fused forms like `im`/`ive`/`dont` now collapse into existing tokens like `i`, `have`, `not` instead of surviving as distinct one-off tokens). Word check on 2,068 rows: `feeling` → `feel` when verb-tagged, stays `feeling` when noun-tagged (both correct). `felt` → `felt` — WordNet gap confirmed.

## 2.6 Stopword removal

Remove common low-signal words **after** lemmatisation using a plain Python set — not `TfidfVectorizer` or any automated vectoriser.
Assignment requires features to be built from tokens manually; no vectoriser should touch the pipeline here.

> Use `sklearn.feature_extraction.text.ENGLISH_STOP_WORDS` or `nltk.corpus.stopwords.words('english')` as your set.
> Explicitly remove negations (`not`, `no`, `nor`, `never`) from the stopword set before filtering — they carry direct emotion signal.

This protection is only effective because of the contraction expansion added in 2.1: `dont`, `cant`, `didnt`, etc. are now normalised to `"do not"`, `"can not"`, `"did not"` upstream, so every negated comment reaches this step with a literal `"not"` token — regardless of which surface form the negation originally arrived in. Before that fix, `negation_words` only matched comments that already spelled negation out in full (`"not"`, `"no"`, `"never"`), and silently missed the ~550 rows where negation arrived fused (`dont`, `cant`, ...).

This step is controlled by the `APPLY_STOPWORDS` flag (set in the imports cell) — defaulting to on, but switchable so stopword removal's actual effect can be measured rather than assumed, same as `APPLY_LEMMATIZATION` in 2.5.

In [13]:
stop_words = get_stop_words(protect_negation=PROTECT_NEGATION)

if APPLY_STOPWORDS:
    df["filtered_tokens"] = df['Lemmatized_Tokens'].apply(
        lambda tokens: remove_stopwords(tokens, protect_negation=PROTECT_NEGATION)
    )
else:
    df["filtered_tokens"] = df['Lemmatized_Tokens']

mask = df['Lemmatized_Tokens'].apply(lambda tokens: any(word in tokens for word in NEGATION_WORDS))

matches = df[mask]
print(f"Rows containing a negation word: {len(matches)}")

for idx in matches.index[:5]:
    original = df.loc[idx, 'Lemmatized_Tokens']
    filtered = df.loc[idx, 'filtered_tokens']

    print(f"Row {idx}:")
    print(f"  original:  {original}")
    print(f"  filtered:  {filtered}")

Rows containing a negation word: 1399
Row 2:
  original:  ['i', 'sit', 'here', 'to', 'write', 'i', 'start', 'to', 'dig', 'out', 'my', 'feeling', 'and', 'i', 'think', 'that', 'i', 'be', 'afraid', 'to', 'accept', 'the', 'possibility', 'that', 'he', 'might', 'not', 'make', 'it']
  filtered:  ['sit', 'write', 'start', 'dig', 'feeling', 'think', 'afraid', 'accept', 'possibility', 'might', 'not', 'make']
Row 4:
  original:  ['i', 'feel', 'suspicious', 'if', 'there', 'be', 'no', 'one', 'outside', 'like', 'the', 'rapture', 'have', 'happen', 'or', 'something']
  filtered:  ['feel', 'suspicious', 'no', 'one', 'outside', 'like', 'rapture', 'happen', 'something']
Row 10:
  original:  ['i', 'feel', 'a', 'bit', 'like', 'franz', 'liebkind', 'in', 'the', 'producer', 'not', 'many', 'people', 'know', 'it', 'but', 'the', 'fuhrer', 'be', 'a', 'terrific', 'dancer']
  filtered:  ['feel', 'bit', 'like', 'franz', 'liebkind', 'producer', 'not', 'many', 'people', 'know', 'fuhrer', 'terrific', 'dancer']
Row 11:


**Result:** 1,399 rows contain at least one preserved negation — up from 932 before the 2.1 contraction fix, now that fused forms like `dont`/`cant`/`didnt` are also normalised to `not` before reaching this step. Row 13 — *"i do not always find myself feel thankful"* — filtered to `['not', 'always', 'find', 'feel', 'thankful', ...]`. Without `not`, those tokens are indistinguishable from a genuinely thankful sentence: `find`, `feel`, `thankful`, `grateful`, `thanks` all present with no signal the sentence is negated. The person is expressing the *absence* of thankfulness — keeping `not` is what makes that classifiable.

### 2.6a Verification: contraction & negation handling

Run the full pipeline (2.1 → 2.6) on a handful of standalone sentences — not from the dataset — to confirm contractions and negation survive end-to-end, the same way 2.5a verified lemmatisation independently before trusting it on the full column.

In [14]:
demo_sentences = [
    "I don't feel happy",
    "I didn't enjoy this",
    "I'm terrified",
    "I can't stop worrying",
]

for sentence in demo_sentences:
    normalized = normalize(sentence)
    filtered = preprocess(sentence, APPLY_LEMMATIZATION, APPLY_STOPWORDS, PROTECT_NEGATION)
    print(f"{sentence!r}")
    print(f"  normalized: {normalized!r}")
    print(f"  filtered:   {filtered}")


"I don't feel happy"
  normalized: 'i do not feel happy'
  filtered:   ['not', 'feel', 'happy']
"I didn't enjoy this"
  normalized: 'i did not enjoy this'
  filtered:   ['not', 'enjoy']
"I'm terrified"
  normalized: 'i am terrified'
  filtered:   ['terrify']
"I can't stop worrying"
  normalized: 'i can not stop worrying'
  filtered:   ['not', 'stop', 'worrying']


**Result:**
```
'I don't feel happy'
  normalized: 'i do not feel happy'
  filtered:   ['not', 'feel', 'happy']
'I didn't enjoy this'
  normalized: 'i did not enjoy this'
  filtered:   ['not', 'enjoy']
"I'm terrified"
  normalized: 'i am terrified'
  filtered:   ['terrify']
"I can't stop worrying"
  normalized: 'i can not stop worrying'
  filtered:   ['not', 'stop', 'worrying']
```

Negation survives in all three applicable cases (`don't`, `didn't`, `can't` all consolidate to `not`). `"terrified"` correctly lemmatises to `"terrify"` (verb-tagged, regular form). `"worrying"` does *not* reduce to `"worry"` here — the tagger labels it `NN` (noun) rather than `VBG` in this short, context-free sentence, so the lemmatiser (correctly, given that tag) leaves it as a noun form. This is the same class of tagger-dependent limitation already documented in 2.5 for irregular verbs — not a regression introduced by this fix, but worth knowing: short, fragment-like inputs give the POS tagger less context, which can occasionally mislabel a word and change how it lemmatises.

## 2.7 Flag and review low-token-count rows

After all transformations, some comments may have 0 or 1 token left. Low token count is used here only to **surface candidates for review** — it is not, by itself, a reason to drop a row. A comment like `"terrified"` would also hit `len(filtered_tokens) <= 1`, but it's a complete, valid, emotion-bearing example, not a defect. The two have to be told apart by reading the raw `Comment`, not by counting tokens.

So this step has two separate parts: (1) flag candidates by token count, (2) individually judge each one against its *raw* comment and only then decide whether it's dropped — with the specific reason recorded per row, not inferred from the flagging threshold itself.

In [15]:
mask = df['filtered_tokens'].apply(lambda tokens: len(tokens) <= 1)

flagged = df[mask]
print(f"Candidate rows with 0 or 1 tokens remaining: {len(flagged)}")
for idx in flagged.index:
    print(f"Row {idx}: filtered_tokens={df.loc[idx, 'filtered_tokens']}  |  raw: {df.loc[idx, 'Comment']!r}"
    )

Candidate rows with 0 or 1 tokens remaining: 6
Row 309: filtered_tokens=['sadness']  |  raw: 'as in sadness a'
Row 896: filtered_tokens=['army']  |  raw: 'in the army'
Row 1291: filtered_tokens=['exam']  |  raw: 'before an exam which i'
Row 3030: filtered_tokens=['sweden']  |  raw: 'in sweden'
Row 4740: filtered_tokens=['school']  |  raw: 'at school'
Row 5363: filtered_tokens=['lecture']  |  raw: 'during lectures'


### 2.7a Per-row review

Reading each candidate's raw comment, not just its surviving token(s):

| Row | Raw comment | Judgment |
|---|---|---|
| 309 | "as in sadness a" | Truncated mid-sentence — trails into nothing. `sadness` survives only because of where the truncation happened, not as a deliberate one-word answer. |
| 896 | "in the army" | Bare context phrase, no emotional content — reads as a clause fragment missing its emotional half. |
| 1291 | "before an exam which i" | Cuts off mid-clause ("which i..."). Incomplete. |
| 3030 | "in sweden" | Bare location phrase, no emotional content. |
| 4740 | "at school" | Bare location phrase, no emotional content. |
| 5363 | "during lectures" | Bare temporal phrase, no emotional content. |

All 6 are genuine truncated fragments on inspection of the raw text — none is a short-but-complete emotional utterance that the length check would have wrongly caught. The outcome happens to match what a blanket rule would have done here, but the *justification* is now the content of each row, not its token count — a future row that was short because it's genuinely a one-word emotional comment would not be dropped by this process.

In [16]:
# Reviewed in 2.7a: all 6 are truncated fragments with no emotional content, not valid
# short emotional utterances. Dropping by explicit index list (not by re-deriving `mask`)
# so this removal is tied to the review above, not to the flagging threshold itself.
fragments_to_drop = [309, 896, 1291, 3030, 4740, 5363]

df = df.drop(index=fragments_to_drop)
df = df.reset_index(drop=True)
print(f"Rows after dropping reviewed fragments: {len(df)}")

Rows after dropping reviewed fragments: 5925


**Result:** 6 rows reviewed individually and dropped — all confirmed truncated sentence fragments with no emotional content (see 2.7a). 5,925 rows remain. This is the same row count as a blanket `len(tokens) <= 1` rule would have produced, but for a different and more defensible reason: each row was checked against its raw text, not assumed problematic from its token count alone.

## 2.8 Label encoding

Convert string labels to integers. Save the encoder to preserve the class mapping for decoding predictions at inference time.

> Print the mapping explicitly: `dict(enumerate(le.classes_))`

In [17]:
le = LabelEncoder()
df['Label'] = le.fit_transform(df['Emotion'])
print(dict(enumerate(le.classes_)))

{0: 'anger', 1: 'fear', 2: 'joy'}


**Result:** Labels encoded as `{0: 'anger', 1: 'fear', 2: 'joy'}`.

## 2.9 Train / validation / test split

Stratified 70 / 15 / 15 split to preserve class proportions in all three sets.

In [18]:
X_train, X_temp, y_train, y_temp = train_test_split(df[['Comment', 'filtered_tokens']], df['Label'], test_size=0.3, train_size=0.7, random_state=42, shuffle=True, stratify=df['Label'])
X_valid, X_test, y_valid, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42, shuffle=True, stratify=y_temp)

**Result:** 5,925 rows split 70/15/15 into train (4,147), validation (889), and test (889) sets.

## 2.10 Verify split balance

Confirm class ratios held across all three splits — a failed stratification silently skews everything downstream.

In [19]:
print(f"train set proportion: {y_train.value_counts(normalize=True) * 100}")
print(f"validation set proportion: {y_valid.value_counts(normalize=True) * 100}")
print(f"test set proportion: {y_test.value_counts(normalize=True) * 100}")

train set proportion: Label
2    33.735230
0    33.662889
1    32.601881
Name: proportion, dtype: float64
validation set proportion: Label
2    33.745782
0    33.633296
1    32.620922
Name: proportion, dtype: float64
test set proportion: Label
2    33.745782
0    33.633296
1    32.620922
Name: proportion, dtype: float64


**Result:** Class proportions hold across all three splits — joy ~33.7%, anger ~33.6%, fear ~32.6% in train, validation, and test alike. Stratification worked as intended.

Validation and test proportions match to 6 decimal places since both came from an even 50/50 split of the same 889-row temp set — not a bug, just two equal-sized stratified halves of the same data.

## 2.11 Save

Persist the processed dataframe and label encoder so Notebook 03 (feature engineering) loads clean, split data without re-running this pipeline.

**Why `assign()` to recombine X/y:** `X_train`/`y_train` were kept separate through the split because `train_test_split` and sklearn's `.fit()` convention expect that shape. But Notebook 03 needs one coherent table per split — to browse rows, filter, and build features from `Comment` + `filtered_tokens` + `Label` together. `assign()` recombines them at the save boundary, matching the shape the next consumer actually wants, without disturbing how the split itself was done.

**Why separate `.pkl` files instead of one combined file:** Downstream steps rarely need every split at once — feature engineering and tuning loop over train/val repeatedly while test stays untouched until final evaluation. Separate files (`train.pkl`, `valid.pkl`, `test.pkl`, `label_encoder.pkl`) let each stage load only what it needs, make it obvious which artifact changed if preprocessing is rerun, and match the convention most ML repos use — versus a single dict-of-everything pickle, which forces loading everything even when only one piece is needed.

In [20]:
train_df = X_train.assign(Label=y_train)
valid_df = X_valid.assign(Label=y_valid)
test_df = X_test.assign(Label=y_test)

In [21]:
to_save = {
    'train.pkl': train_df,
    'valid.pkl': valid_df,
    'test.pkl': test_df,
    'label_encoder.pkl': le,
}
for filename, obj in to_save.items():
    with open(f'../data/processed/{filename}', 'wb') as f:
        pickle.dump(obj, f)

**Result:** Saved `train.pkl` (4,147 rows), `valid.pkl` (889 rows), `test.pkl` (889 rows), and `label_encoder.pkl` to `../data/processed/`. Each split table carries `Comment`, `filtered_tokens`, and `Label` together.